### PR TITLE
Project Enumerable methods

### PR DESCRIPTION
**Pull request by petudeveloper**
Following the git-flow workflow:

1. Created the development branch as a copy for the main branch.
2. We created Enumerable Methods branch to serve as our code implementation for the whole process. 
3. We set up rubocop to fix linters errors and create our gem file.
4. We created rb file for our Enumerable Methods implementation.
5. We created a separate test file to test the different case scenarios for the methods.
6. We included proper documentation for the project.

### CODE REVIEW

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hello, you have been doing a tremendous job so far :ok_hand:
## But there are some changes to make in order to get your project approved.

- [x]  Please remove commented code because it makes the overall code unclean.
### #my_each
- [x] It should not use #each in implementation. Please see [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png)
### #my_each_with_index
- [x] It should return an array when called on an array, but now it's returning a range. It shows an error when called on a range.
### #my_select
- [x] Please see #select in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select) and make #my_select works same.
- [x] It should return an enumerator if no block is given.
### #my_all?
- [x] Please see #all? in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F) and make #my_all? works same.
- [x] It should return a boolean when no block is given and a Regex is passed as an argument.
- [x] It should return a boolean when no block is given and a Class is passed as an argument.
- [x] It should return false when no block is given and one of the collection members is false or nil.
- [x] It should return true when no block is given and none of the collection members are false or nil.
### #my_any?
- [x] Please see #any? in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F) and make #my_any? works same.
- [x] It should return a boolean when no block is given and a Regex is passed as an argument.
- [x] It should return a boolean when no block is given and a Class is passed as an argument.
- [x] It should return false when no block is given and at least one of the collections is not true.
- [x] It should return true when no block is given and at least one of the collections is not false or nil.
### #my_none?
- [x] Please see #none? in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F) and make #my_none? works same.
- [x] It returns false if the block ever returns true for all elements
- [x] It should return a boolean when no block is given and a Regex is passed as an argument.
- [x] It should return false when no block is given and one of the collection members is true.
- [x] It should return true when no block is given and none of the collection members is true.
- [x] It should return true when no block or argument is passed and called on an empty array. 
### #my_count?
- [x] Please see #count? in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count) and make #my_count? works same.
- [x] It should count the elements when called on a Range.
- [x] It should count the number of items in an array that are equal to the item if an argument is given.
- [x] It should count the number of items in a Range that are equal to the item if an argument is given.
- [x] It should count the number of items in a Hash that are equal to the item if an argument is given.
### #my_map?
- [x] Please see #map? in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map) and make #my_map? works same.
- [x] It should return an Enumerator if no block is given
### #my_inject?
- [x] Please see #inject? in [Ruby docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject) and make #my_inject? works same.
- [x] It should combine each element of the collection by applying the symbol as a named method when a symbol is specified without an initial value.
- [x]  It should combine each element of the collection by applying the symbol as a named method when a symbol is specified with an initial value.
- [x] It should combine all elements of enum by applying a binary operation, specified by a block::range when a block is given with an initial value.

### Happy Coding! 

## You can always connect with me on slack ( @Sajjad Ahmad ).

**Comment by petudeveloper** [comment]
@SajjadAhmad14 Thank you for your comments, they helped enlarge our understanding of the subject matter, and we were able to digest ruby docs more religiously.

**Comment by DanSam5K** [comment]
Absoluetly!👌

**Review by SajjadAhmad14 — CHANGES_REQUESTED** [review]
# Changes Required :recycle:

## You have made most of the changes suggested previously :ok_hand:
## But there are some edge cases you need to work on and your project will be ready to get approved!

- [x] #my_each is returning an array when called on a range and printing nil when called on hash. It should return a Range when called on a Range and should be ok working with a Hash.
- [x] #my_each_with_index is still returning a Range when called on an Array, giving an error when called on a Range and returning a Range when called on a Hash. It should return a Hash when called on a Hash, an Array when called on an Array, and should be working with a Range.
- [x] #my_count is giving an error when called on a Range.

### Happy Coding!

## If you have any questions please connect with me on slack ( @Sajjad Ahmad ).

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coders, 👋🏽 
**This is a great project you have going on, well-done lads and thanks for implementing the changes requested earlier on**. 
### Well-done 💯 
    1. The enumerable methods almost pass all test cases.
    2. The linter checks run just fine 👍🏽 

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

- REQUIRED

- [x] Check the inline comment below.

#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      my_each { |item| result << item if yield(item) }
+      result.my_each_with_index(&counter)
+    else
+      my_each_with_index(&counter)
+    end
+    is_a?(Hash) ? count - 2 : count
+  end
+
+  # 8. my map method
+  def my_map
```
Comment (mke2111) [inline enumerable_methods.rb]: ### The `my_map` method should only execute the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

An argument error is raised when I pass an argument to the enumerable method.
```
Failure/Error: expect(array.my_map(my_proc) { |num| num < 10 }).to eq array.map(&my_proc)

ArgumentError:
       wrong number of arguments (given 1, expected 0)

```
> Modify this method to accept at least one argument passed to it.

Comment (petudeveloper) [reply enumerable_methods.rb]: Thanks for the feedback. We implemented the suggested changes.

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details open="open">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] `#my_map` returns an Enumerator if no block is given
- [ ] `#my_map` executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.



</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [x] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Comment by DanSam5K** [comment]
Thank you @jubaan for providing a clearer view of what needs to be done on slack🙏🏼

**Comment by DanSam5K** [comment]
Thank you @jubaan for providing a clearer view of what needs to be done on slack🙏🏼

**Review by mricanho — APPROVED** [review]
# Status: ✔️ Project Approved ✔️ 

Hello @DanSam5K @petumazo, good job on your enumerables! 🔥 

You worked hard in all the instructions that the previous Code Reviewers give you👍🏽 

Well done team


Now you can merge this branch 🚀 

You can contact me on slack if you have any questions, just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!  👨🏽‍💻 🖥️

**Comment by DanSam5K** [comment]
## @mricanho 😅 
- Thank you for Empathizing 🤝

**Pull request merged**